### PR TITLE
Implement configuration file support

### LINKED ISSUE #502: Add option for configuration file
Currently all settings need to be passed as options on the command line. We should add an option to use a configuration file. I suggest to look into how pip handles that, as it allows all command line options to be supplied via a config file.

We should also look into whether we should support default locations or if it exclusively works via the command line option.

There currently is some mix of state and configuration kept in ``.nodeinfo`` which must also be handled differently.

This change is needed to remove state from the filesystem to fully implement #373

### PR DESCRIPTION
**Pull request by fschulze**
Fixes #502

### CODE REVIEW

File: client/testing/test_quickstart.py [hunk 1]
```diff
@@ -10,7 +10,7 @@ def test_dryrun(cmd_devpi):
 
 @pytest.mark.skipif("config.option.fast")
 def test_functional(cmd_devpi, monkeypatch, tmpdir):
-    monkeypatch.setenv("DEVPI_SERVERDIR", tmpdir.join("server"))
+    monkeypatch.setenv("DEVPISERVER_SERVERDIR", tmpdir.join("server"))
     monkeypatch.setenv("DEVPI_CLIENTDIR", tmpdir.join("client"))
```
Comment (gotcha) [inline client/testing/test_quickstart.py]: why is this not changed to ``DEVPICLIENT_CLIENTDIR``?

File: doc/Makefile [hunk 2]
```diff
@@ -30,7 +30,7 @@ USER_MAN_CHAPTERS = commands\
           
                                  
 #export DEVPI_CLIENTDIR=$(CURDIR)/.tmp_devpi_user_man/client
```
Comment (gotcha) [inline doc/Makefile]: why is this not changed to ``DEVPICLIENT_CLIENTDIR`` ?

File: server/devpi_server/config.py [hunk 3]
```diff
@@ -240,52 +242,143 @@ def try_argcomplete(parser):
...
+    config_dirs = config_dirs.split(os.pathsep)
+    config_dirs.append(
+        appdirs.user_config_dir('devpi-server', 'devpi'))
+    config_files = []
+    for config_dir in config_dirs:
+        config_file = os.path.join(config_dir, 'devpi-server.yml')
+        if os.path.exists(config_file):
+            config_files.append(config_file)
+    if len(config_files) > 1:
+        log.warn("Multiple configuration files found:\n%s", "\n".join(config_files))
```
Comment (gotcha) [inline server/devpi_server/config.py]: I would mention which file will be actually used.

File: server/devpi_server/config.py [hunk 4]
```diff
@@ -240,52 +242,143 @@ def try_argcomplete(parser):
...
+            return {}
+        elif config.is_sequence():
+            raise InvalidConfigError("The config file must be a mapping, not a sequence.")
+        if 'devpi-server' not in config:
+            return {}
+        config = config['devpi-server']
+        if config.is_scalar():
+            return {}
+        elif config.is_sequence():
+            raise InvalidConfigError("The config file must be a mapping, not a sequence.")
```
Comment (gotcha) [inline server/devpi_server/config.py]: Same message as just before ?

File: server/devpi_server/config.py [hunk 5]
```diff
@@ -240,52 +242,143 @@ def try_argcomplete(parser):
...
+    parser.post_process_actions(defaultget=defaultget)
+    args = parser.parse_args(argv[1:])
+    if args.help is True:
+        parser.print_help()
+        parser.exit()
     config = Config(args, pluginmanager=pluginmanager)
     return config
 
+
+def get_action_long_name(action):
```
Comment (gotcha) [inline server/devpi_server/config.py]: add a comment where this comes from

File: server/devpi_server/config.py [hunk 6]
```diff
@@ -240,52 +242,143 @@ def try_argcomplete(parser):
...
-        return opt
-
-    def _processopt(self, opt):
-        try:
-            opt.default = self._defaultget(opt.dest)
-        except KeyError:
-            pass
-        if opt.help and opt.default:
-            opt.help += " [%s]" % opt.default
+    def post_process_actions(self, defaultget=None):
```
Comment (gotcha) [inline server/devpi_server/config.py]: add a comment about what it does

File: doc/quickstart-server.rst [hunk 7]
```diff
@@ -38,6 +38,30 @@ And let's check the version::
...
+
+    devpi-server:
+      serverdir: /var/db/devpi-server
+      host: localhost
+      port: 8080
+
+.. _`strict YAML`: https://pypi.org/project/strictyaml/
+.. _`appdirs`: https://pypi.org/project/appdirs/
+
+
```
Comment (gotcha) [inline doc/quickstart-server.rst]: no mention of environment variables

File: client/testing/test_quickstart.py (hunk 1 above)
Comment (fschulze) [reply client/testing/test_quickstart.py]: This PR only adds configuration file support to devpi-server. If we ever add it to devpi-client, then it should be updated there.

**Pull request merged**